server/internal/mcp/rpc_tools_list.go

```diff
@@ -83,3 +80,216 @@ func handleToolsList(ctx context.Context, logger *slog.Logger, db *pgxpool.Pool,
 
 	return bs, nil
 }
+
+func buildToolListEntries(tools []*types.Tool) []*toolListEntry {
+	result := make([]*toolListEntry, 0, len(tools))
+	for _, tool := range tools {
+		if entry := toolToListEntry(tool); entry != nil {
+			result = append(result, entry)
+		}
+	}
+	return result
+}
+
+func toolToListEntry(tool *types.Tool) *toolListEntry {
+	if tool == nil {
+		return nil
+	}
+
+	var meta map[string]any
+	if tool.FunctionToolDefinition != nil {
+		meta = tool.FunctionToolDefinition.Meta
+	}
+
+	baseTool := conv.ToBaseTool(tool)
+
+	return &toolListEntry{
+		Name:        baseTool.Name,
+		Description: baseTool.Description,
+		InputSchema: json.RawMessage(baseTool.Schema),
+		Meta:        meta,
+	}
+}
+
+func buildProgressiveSessionTools(toolset *types.Toolset) []*toolListEntry {
```

Comment:
I'd move this stuff into its own file for progressive_tool_calling

Reply:
good idea